### PR TITLE
4844: blob encoding version 2

### PR DESCRIPTION
**Pull request by angel-ding-cb**
**Description**

This is the implementation based on the 2nd version of [blob encoding spec](https://github.com/ethereum-optimism/optimism/pull/8745/files?short_path=882757e#diff-882757e7a296e749f616fb3e5972b2f8041dc9c515876529981fddd38e73a6a7). 

**Tests**

Tested small and large data inputs, as well as edge cases.

**Metadata**

- Specs #[[Link to Spec](https://github.com/ethereum-optimism/optimism/pull/8745/files?short_path=882757e#diff-882757e7a296e749f616fb3e5972b2f8041dc9c515876529981fddd38e73a6a7)]

### CODE REVIEW

File: op-service/eth/blob_test.go [hunk 1]
```diff
@@ -32,10 +33,63 @@ func TestBlobEncodeDecode(t *testing.T) {
 	}
 }
 
+func TestSmallBlobEncoding(t *testing.T) {
+	// the first field element is filled and no data remains
+	data := Data(make([]byte, 128))
+	data[127] = 0xFF
+
```
Comment (tynes) [inline op-service/eth/blob_test.go:37]: We usually use `"github.com/stretchr/testify/require"` for test assertions.

You can do `require.NoError(b.FromData(data), "failed to encode bytes")`

File: op-service/eth/blob.go [hunk 2]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
-// bytes.
-//
-// The remaining field elements each encode 31 bytes of the remaining input data, up until the end
-// of the input.
-//
-// TODO: version the encoding format to allow for future encoding changes
+// It reads 31 bytes and then 1 byte for three times and then 31 bytes of data from the input.
+// For all the 31 bytes of data, they are encoded into [1:32] bytes of each field element.
+// For the extra 3 bytes read from the input, they are encoded into the 1 bytes of data on the top of each field element.
+// So for each field (4 field elements), 3 bytes are encoded into 4 bytes where the highest order bit is set to 0
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]:
```suggestion
//
// In each round we perform 7 reads of input of lengths (31,1,31,1,31,1,31) bytes respectively for a total of 127
// bytes. This data is encoded into the next 4 field elements of the output by placing each of the 4x31
// byte chunks into bytes [1:32] of its respective field element. The three single byte chunks (24 bits) are
// split into 4x6-bit chunks, each of which is written into the top most byte of its respective field element,
// leaving the top 2 bits of each field element empty to avoid modulus overflow.
```

File: op-service/eth/blob.go [hunk 3]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+
+	// first field element encodes the version and the length of the data in [1:5]
+	b[1] = EncodingVersion
+
+	// encode the length as big-endian uint24 into [2:5] bytes of the first field element
+	if len(data) < 1<<24 {
+		// Zero out any trailing data in the buffer if any
+		b[2] = byte((len(data) >> 16) & 0xFF) // Most significant byte
+		b[3] = byte((len(data) >> 8) & 0xFF)
+		b[4] = byte(len(data) & 0xFF) // Least significant byte
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go:94]: Why not use what was there before? It's more concise & probably more efficient.
```suggestion
	binary.LittleEndian.PutUint32(b[1:5], uint32(len(data)))
```

File: op-service/eth/blob.go [hunk 4]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+	} else {
+		return fmt.Errorf("Error: length_rollup_data is too large")
+	}
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]: It's more idiomatic in go to nest only the exceptional case, e.g. (with some other edits/simplifications for clarity): 
```suggestion
    ilen := uint32(len(data))
    if (ilen >> 24) != 0 {
        return fmt.Errorf("unexpected bits in top byte of length field. len=%v", ilen)
    }
    b[2] = byte(ilen >> 16)
    b[3] = byte(ilen >> 8)
    b[4] = byte(ilen)
```

File: op-service/eth/blob.go [hunk 5]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+	buffer, offset = read(offset, 31, data)
+	encodedByte = ((z[0] & 0b1100_0000) >> 2) | ((y[0] & 0b1111_0000) >> 4)
+	pointer = b.write([]byte{encodedByte}, pointer)
+	pointer = b.write(buffer, pointer)
+
+	if offset == len(data) {
+		return nil
+	}
+
+	for fieldNumber := 1; fieldNumber < 1024; fieldNumber++ {
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]:
```suggestion
	for round := 1; round < 1024; round++ {
```
You might also define a "NumRounds" constant with value 1024 and use this here and in the decoder too.

File: op-service/eth/blob.go [hunk 6]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
 	if offset < len(data) {
 		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-offset)
 	}
+
 	return nil
 }
 
+func read(offset int, numBytes int, data []byte) ([]byte, int) {
+	if offset >= len(data) {
+		// If the offset is at or beyond the end of data, return a new byte array of numBytes length
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]: cute, I was wondering how you were dealing with short inputs... I would put this detail about how short reads in the godoc above the function as it's important to understanding how the encoding works. 

Something like "read(..) always returns an array of length exactly numBytes, appropriately padded with zeros should there be not enough input"

File: op-service/eth/blob.go [hunk 7]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+		b[3] = byte((len(data) >> 8) & 0xFF)
+		b[4] = byte(len(data) & 0xFF) // Least significant byte
+	} else {
+		return fmt.Errorf("Error: length_rollup_data is too large")
+	}
+
+	offset := 0
+	var buffer []byte
+
+	// encode the first 27 + 1 bytes of data into remaining bytes of first field element
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]:
```suggestion
	// round 0 is special cased here since we can only encode the first (31-5)+1=27 bytes of input into
	//  the first field element due to version/length encoding already occupying the first 5 bytes.
```

File: op-service/eth/blob.go [hunk 8]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
-		if b[i*32] != 0 {
-			return nil, fmt.Errorf("invalid blob, found non-zero high order byte %x of field element %d", b[i*32], i)
-		}
-		copy(data[i*31:i*31+31], b[i*32+1:i*32+32])
+	data := make(Data, BlobSize)
+	firstField := b[:FieldSize]
+
+	// check the version
+	if firstField[1] != EncodingVersion {
+		return nil, fmt.Errorf("invalid blob, expected version %d, got %d", EncodingVersion, firstField[0])
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]:
```suggestion
		return nil, fmt.Errorf("invalid blob, expected version %d, got %d", EncodingVersion, firstField[1])
```

File: op-service/eth/blob.go [hunk 9]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
 	}
-	// extract the length prefix & trim the output accordingly
-	dataLen := binary.LittleEndian.Uint32(data[:4])
-	data = data[4:]
-	if dataLen > uint32(len(data)) {
+
+	// decode the 3-byte length prefix into 4-byte integer
+	var dataLen int32
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]:
```suggestion
	var dataLen uint32
```

File: op-service/eth/blob.go [hunk 10]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+
+	// Assuming b[2], b[3], and b[4] contain the encoded length in big-endian format
+	dataLen = int32(b[2]) << 16 // Shift the most significant byte 16 bits to the left
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]: int32 -> uint32 here & below

File: op-service/eth/blob.go [hunk 11]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+
+	// encode the first 27 + 1 bytes of data into remaining bytes of first field element
+	buffer, offset = read(0, 27, data)
+	x, offset := read(offset, 1, data)
+	encodedByte := x[0] & 0b0011_1111
+	b.write([]byte{encodedByte}, 0)
+	b.write(buffer, 5)
+	pointer := 32 // manually set the pointer to the next field element
+
+	clearBuffer(buffer)
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]: I don't think you need these clearBuffer calls because read() always returns a new array.

File: op-service/eth/blob.go [hunk 12]
```diff
@@ -131,3 +272,9 @@ func (b *Blob) Clear() {
 		b[i] = 0
 	}
 }
+
+func clearBuffer(buffer []byte) {
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]: you can get rid of this once you get rid of the clearBuffer calls as suggested above.

File: op-service/eth/blob.go [hunk 13]
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+
+	// Calculate the actual number of bytes to read, which may be less than numBytes
+	// if the offset is near the end of data
+	actualNumBytes := numBytes
+	if offset+numBytes > len(data) {
+		actualNumBytes = len(data) - offset
+	}
+
+	// Create a new byte array and copy the data from the original slice
+	byteArray := make([]byte, actualNumBytes)
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go]: Preserve the invariant that the returned byte array is always exactly "numBytes".  While it works either way, I think this makes it easier to reason about.  You could even move the byteArray creation to the very top of the function to make this exceptionally clear, and return it even in the offest >= len(data) case.

```suggestion
	byteArray := make([]byte, numBytes)
```

File: op-service/eth/blob_test.go (hunk 1 above)
Comment (roberto-bayardo) [reply op-service/eth/blob_test.go:37]: updated to use testify/require

Comment (tynes) [reply op-service/eth/blob_test.go:37]: thank you!

File: op-service/eth/blob_test.go [hunk 14]
```diff
@@ -54,16 +94,20 @@ func TestInvalidBlobDecoding(t *testing.T) {
...
+	b[32] = 0x0
 
-	b[32] = 0x00
-	b[4] = 0xFF // encode an invalid (much too long) length prefix
-	if _, err := b.ToData(); err == nil {
-		t.Errorf("expected error, got none")
-	}
+	b[VersionOffset] = 0x01 // invalid encoding version
+	_, err = b.ToData()
+	require.Error(t, err)
```
Comment (tynes) [inline op-service/eth/blob_test.go]: Can we use the following method to assert specifically on certain errors? https://pkg.go.dev/github.com/stretchr/testify/require#ErrorIs

If we break out the errors into their own types then this becomes much easier. We should have test coverage of each error case

File: op-service/eth/blob.go [hunk 15]
```diff
@@ -70,60 +72,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+	output = output[:outputLen]
+	return output, nil
+}
+
+// decodeFieldElement decodes the next input field element by writing its lower 31 bytes into its
+// appropriate place in the output and checking the high order byte is valid.
+func (b *Blob) decodeFieldElement(opos, ipos int, encodedByte []byte, output []byte) (int, int, error) {
+	// two highest order bits of the first byte of each field element should always be 0
+	if b[ipos]&(0xC0) != 0 {
+		return 0, 0, fmt.Errorf("invalid blob, field element %d has a high order bit set", ipos)
```
Comment (tynes) [inline op-service/eth/blob.go]: I do not see any test coverage of this error

File: op-service/eth/blob_test.go [hunk 16]
```diff
@@ -1,7 +1,10 @@
 package eth
 
 import (
+	"math/rand"
 	"testing"
+
+	"github.com/stretchr/testify/require"
 )
```
Comment (tynes) [inline op-service/eth/blob_test.go:9]: It would be nice to have some fuzz roundtrip tests for encode/decode

File: op-service/eth/blob.go [hunk 17]
```diff
@@ -70,60 +72,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
-// bytes.
-//
-// The remaining field elements each encode 31 bytes of the remaining input data, up until the end
-// of the input.
-//
-// TODO: version the encoding format to allow for future encoding changes
+// For only the very first output field, bytes [1:5] are used to encode the version and the length
+// of the data.
 func (b *Blob) FromData(data Data) error {
 	if len(data) > MaxBlobDataSize {
```
Comment (tynes) [inline op-service/eth/blob.go:94]: There are 2 checks for the size, are both necessary?

Comment (tynes) [reply op-service/eth/blob.go:94]: Both error messages are the same so its not possible to distinguish a difference between the 2 cases. If both are required, we should modify the error messages so they are distinct and then test for both cases

File: op-service/eth/blob.go (hunk 15 above)
Comment (roberto-bayardo) [reply op-service/eth/blob.go]: see first case of TestInvalidBlobDecoding

File: op-service/eth/blob.go (hunk 17 above)
Comment (roberto-bayardo) [reply op-service/eth/blob.go:94]: the second isn't necessary, was more for documentation that the high order byte will never be set and we can always squeeze this into 3 bytes. removed & replaced with a comment.

File: op-service/eth/blob_test.go (hunk 16 above)
Comment (roberto-bayardo) [reply op-service/eth/blob_test.go:9]: added

File: op-service/eth/blob_test.go (hunk 14 above)
Comment (roberto-bayardo) [reply op-service/eth/blob_test.go]: done

File: op-service/eth/blob.go [hunk 18]
```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
-	data = data[:dataLen]
-	return data, nil
+	copy(output[opos:], b[ipos+1:ipos+32])
+	return opos + 32, ipos + 32, nil
+}
+
+// reassembleBytes takes the 4x6-bit chunks from encodedByte, reassembles them into 3 bytes of
+// output, and places them in their appropriate output positions.
+func reassembleBytes(opos int, encodedByte []byte, output []byte) int {
+	opos-- // account for fact that we don't output a 128th byte
```
Comment (tynes) [inline op-service/eth/blob.go:260]: Should we be worried about underflow here?

File: op-service/eth/blob_test.go [hunk 19]
```diff
@@ -72,7 +116,18 @@ func TestTooLongDataEncoding(t *testing.T) {
 	data := Data(make([]byte, BlobSize))
 	var b Blob
 	err := b.FromData(data)
-	if err == nil {
-		t.Errorf("expected error, got none")
-	}
+	require.Error(t, err)
+}
+
+func FuzzEncodeDecodeBlob(f *testing.F) {
```
Comment (tynes) [inline op-service/eth/blob_test.go:122]: You will need to add a call to run this fuzz test in the makefile under `fuzz`

https://github.com/ethereum-optimism/optimism/blob/develop/op-service/Makefile

File: op-service/eth/blob.go [hunk 20]
```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
 //
-// The remaining field elements each encode 31 bytes of the remaining input data, up until the end
-// of the input.
-//
-// TODO: version the encoding format to allow for future encoding changes
+// For only the very first output field, bytes [1:5] are used to encode the version and the length
+// of the data.
 func (b *Blob) FromData(data Data) error {
 	if len(data) > MaxBlobDataSize {
 		return fmt.Errorf("data is too large for blob. len=%v", len(data))
```
Comment (tynes) [inline op-service/eth/blob.go]: I am sorry if this is annoying but I don't see any test coverage of this error case

File: op-service/eth/blob.go [hunk 21]
```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+
+		buffer, ipos = read(ipos, 31, data)
+		encodedByte = ((z[0] & 0b1100_0000) >> 2) | ((y[0] & 0b1111_0000) >> 4)
+		opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
 	}
-	if offset < len(data) {
-		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-offset)
+
+	if ipos < len(data) {
+		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-ipos)
```
Comment (tynes) [inline op-service/eth/blob.go]: Is there test coverage of this error case?

File: op-service/eth/blob.go [hunk 22]
```diff
@@ -14,7 +14,16 @@ import (
...
 	BlobSize        = 4096 * 32
-	MaxBlobDataSize = 4096*31 - 4
+	MaxBlobDataSize = (4*31+3)*1024 - 4
+	EncodingVersion = 0
+	VersionOffset   = 1    // offset of the version byte in the blob encoding
+	Rounds          = 1024 // number of encode/decode rounds
+)
+
+var (
+	InvalidFieldElementError    = errors.New("invalid field element")
```
Comment (tynes) [inline op-service/eth/blob.go]: nit: errors are usually prefixed with `Err` and we may want to prefix with `ErrBlob` because this is in the `eth` package

File: op-service/eth/blob.go (hunk 18 above)
Comment (roberto-bayardo) [reply op-service/eth/blob.go:260]: reassembleBytes is only called after processing 4 field elements which advances opos by 128.  Max lookback here is within that window, so unless someone introduces a coding error or starts calling this from elsewhere, underflow can't happen.  Could extend this func to also return error if you want me to add an explicit test just in case.

File: op-service/eth/blob.go (hunk 22 above)
Comment (roberto-bayardo) [reply op-service/eth/blob.go]: added ErrBlob prefix

File: op-service/eth/blob.go (hunk 20 above)
Comment (roberto-bayardo) [reply op-service/eth/blob.go]: no worries :) There is coverage though, see TestTooLongDataEncoding test.   I just updated it to return a specific error type to test for.

File: op-service/eth/blob_test.go (hunk 19 above)
Comment (roberto-bayardo) [reply op-service/eth/blob_test.go:122]: done

File: op-service/Makefile [hunk 23]
```diff
@@ -14,4 +14,6 @@ fuzz:
 	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadUnmarshal ./eth
 	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadMarshalUnmarshalV1 ./eth
 	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadMarshalUnmarshalV2 ./eth
+	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadMarshalUnmarshalV2 ./eth
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply op-service/Makefile]: @tynes is this duplicate test here intentional or should I fix?

File: op-service/eth/blob.go (hunk 21 above)
Comment (roberto-bayardo) [reply op-service/eth/blob.go]: no because it's one of those errors that shouldn't ever trigger, the topmost check should catch all length issues. this is more for catching coding errors if the code is modified.

File: op-service/Makefile (hunk 23 above)
[the remark this answers is not in the document]
Comment (tynes) [reply op-service/Makefile]: The rabbit is correct, do you mind removing it?

[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply op-service/Makefile]: done

**Review by trianglesphere — COMMENTED** [review]
The parsing looks pretty good to me & I like the round trip tests. One class of tests I'd like to see is a one way test or test vectors. i.e. Given a known input data, we should expect it to yield a known output. We should check both ways (i.e. given some data, expect a blob output & given a blob, expect some data), but it's nice to have the specific output which can manually be verified.

**Comment by roberto-bayardo** [comment]
> One class of tests I'd like to see is a one way test or test vectors. i.e. Given a known input data, we should expect it to yield a known output. We should check both ways (i.e. given some data, expect a blob output & given a blob, expect some data), but it's nice to have the specific output which can manually be verified.

OK to add test vectors as a followup PR? I created an issue for it to make sure it doesn't get dropped and added a TODO referencing it in blob_test.go.

**Review by trianglesphere — APPROVED** [review]
Doing 1 way tests is ok as a follow up

File: op-service/eth/blob.go [hunk 24]
```diff
@@ -70,60 +80,193 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
 
+// read reads up to numBytes of input from the data starting at the given offset, and returns an
+// updated offset of max(offset+numBytes, len(data)).  It always returns a byte array of length
+// exactly numBytes, even if there wasn't enough data to read. In the short-read case, any trailing
+// bytes are zero padded.
+func read(offset int, numBytes int, data []byte) ([]byte, int) {
+	byteArray := make([]byte, numBytes)
+	if offset >= len(data) {
+		return byteArray, len(data)
+	}
```
Comment (protolambda) [inline op-service/eth/blob.go]: We should not be allocating a new slice of bytes for every read call.
If we make `read` internal` to the decoding function,  then we can safely avoid the `ipos` input/output, and just return a slice of the original  bytes. This then makes it not allocate so much, and much less verbose. I'll submit a quick PR.

Comment (protolambda) [reply op-service/eth/blob.go]: https://github.com/roberto-bayardo/optimism/pull/10

**Review by protolambda — APPROVED** [review]

**Review by tynes — APPROVED** [review]
